src/reducer/admin/resource/index.js

```diff
@@ -1,9 +1,37 @@
-import { combineReducers } from 'redux';
+import { DECLARE_RESOURCES } from '../../../actions';
+
 import data from './data';
 import list from './list';
 
-export default resource =>
-    combineReducers({
-        data: data(resource),
-        list: list(resource),
-    });
+const initialState = {};
+export default (previousState = initialState, action) => {
+    if (action.type === DECLARE_RESOURCES) {
+        const newState = action.payload.reduce(
+            (acc, resource) => ({
+                ...acc,
+                [resource.name]: {
+                    props: resource,
+                    data: data(resource.name)(undefined, action),
+                    list: list(resource.name)(undefined, action),
+                },
+            }),
+            {}
+        );
+        return newState;
+    }
+
+    const resources = Object.keys(previousState);
+    const newState = resources.reduce(
+        (acc, resource) => ({
+            ...acc,
+            [resource]: {
+                props: previousState[resource].props,
+                data: data(resource)(previousState[resource].data, action),
+                list: list(resource)(previousState[resource].list, action),
+            },
+        }),
+        {}
+    );
+
+    return newState;
```

Comment:
On logout, we reinitialize the state completely. See [here](https://github.com/marmelab/admin-on-rest/pull/959/files#diff-e3f4353cfcf575a2f126ee1e33b9b51fR43) (already the case on master).